src/ptvsd/adapter/ide.py

```diff
@@ -222,58 +260,101 @@ def launch_request(self, request):
         )
         console_title = request("consoleTitle", json.default("Python Debug Console"))
 
-        self.session.spawn_debuggee(request, sudo, args, console, console_title)
-
-        if "RedirectOutput" in self.session.debug_options:
-            # The launcher is doing output redirection, so we don't need the server.
-            request.arguments["debugOptions"].remove("RedirectOutput")
+        launcher.spawn_debuggee(
+            self.session, request, sudo, args, console, console_title
+        )
 
     @_start_message_handler
     def attach_request(self, request):
         if self.session.no_debug:
             raise request.isnt_valid('"noDebug" is not supported for "attach"')
 
-        pid = request("processId", int, optional=True)
-        if pid == ():
-            # When the adapter is spawned by the debug server, it is connected to the
-            # latter from the get go, and "host" and "port" in the "attach" request
-            # are actually the host and port on which the adapter itself was listening,
-            # so we can ignore those.
-            if self.server:
-                return
+        # There are four distinct possibilities here.
+        #
+        # If "processId" is specified, this is attach-by-PID. We need to inject the
+        # debug server into the designated process, and then wait until it connects
+        # back to us. Since the injected server can crash, there must be a timeout.
+        #
+        # If "subProcessId" is specified, this is attach to a known subprocess, likely
+        # in response to a "ptvsd_attach" event. If so, the debug server should be
+        # connected already, and thus the wait timeout is zero.
+        #
+        # If neither are specified, but "listen" is true, this is attach-by-socket
+        # with the server expected to connect to the adapter via ptvsd.attach(). There
+        # is no PID known in advance, so just wait until the first server connection
+        # indefinitely, with no timeout.
```

Comment:
>indefinitely, with no timeout. [](start = 10, length = 30)

Do we need a way to time out?

Reply:
No, this is intentional - the whole point of "listen" is to have an adapter waiting for the server to attach to it, however long that takes.